### PR TITLE
Fix: Correctly delete files using ImageInput 

### PR DESCRIPTION
**Pull request by mguihal**
Hi,

When using the ImageInput component, we can: 
- Add a new file
- Delete a previously uploaded file
- Change the uploaded file with a new one

The last two cases were not correctly handled (files were not deleted neither in the database, nor in the filesystem).

- I changed the value returned by the ImageInput component to add the previous uploaded file too if any, and added the mecanism in the semantic-data-provider to handle file deletion.

- I changed the file type in the middleware deletion action in the LDP service to handle correctly the resource type (related to https://github.com/assemblee-virtuelle/semapps/issues/1185)

*NB: this PR is based on the https://github.com/assemblee-virtuelle/semapps/pull/1196 for facility, but the two are independant*

### CODE REVIEW

File: src/middleware/packages/ldp/services/resource/actions/delete.js [hunk 1]
```diff
@@ -40,7 +40,7 @@ module.exports = {
       await ctx.call('ldp.container.detach', { containerUri, resourceUri, webId: 'system' });
     }
 
-    if (oldData['@type'] === 'semapps:File') {
+    if (oldData['type'] === 'File') {
```
Comment (srosset81) [inline src/middleware/packages/ldp/services/resource/actions/delete.js]: Are you sure ? I don't think the type will ever be `'File'`, except if you use a JSON-LD context of your own.
Currently handling JSON-LD is a mess, which is why we are thinking of using a (custom ?) library for that.

**Review by srosset81 — CHANGES_REQUESTED** [review]
Thanks a lot for these proposals of improvement !
Not deleting files is indeed a problem. However if we did not delete files until now, there was a reason: it is because the handling is made *before* the POST or PUT call with the resource. For the POST/create method it's not a problem (we won't delete any file anyway) but if the PUT/update method fails, the files will be deleted and we won't be able to get them back. The PUT method is particularly sensitive because, if we only have a `acl:Append` right, it will succeed only if we add new data (but not if we delete or overwrite existing data)
A solution could be to delete the files only *after* the PUT returns successfully. At that moment, we can safely delete the files. `handleFiles` could return a list of files to delete, and it could be effectively deleted after the PUT call ?

**Comment by srosset81** [comment]
Another problem I think of: if the same file is being used/linked by other resources, then deleting the file will make it unavailable for those other resources. Currently there is nothing that allows to do this (if you want to add the same image on two documents, you must upload it twice), but this is a possible scenario in the future, isn't it ? For example if we had some kind of digital assets management library like https://chonky.io

This is why I had suggested this idea: https://github.com/assemblee-virtuelle/semapps/issues/1147

File: src/middleware/packages/ldp/services/resource/actions/delete.js (hunk 1 above)
Comment (mguihal) [reply src/middleware/packages/ldp/services/resource/actions/delete.js]: Yes, I updated. Only the @ was incorrect

**Comment by mguihal** [comment]
> Not deleting files is indeed a problem. However if we did not delete files until now, there was a reason: it is because the handling is made before the POST or PUT call with the resource. For the POST/create method it's not a problem (we won't delete any file anyway) but if the PUT/update method fails, the files will be deleted and we won't be able to get them back. The PUT method is particularly sensitive because, if we only have a acl:Append right, it will succeed only if we add new data (but not if we delete or overwrite existing data)
A solution could be to delete the files only after the PUT returns successfully. At that moment, we can safely delete the files. handleFiles could return a list of files to delete, and it could be effectively deleted after the PUT call ?

I updated the PR to split in two functions the upload and deletion of files. One is called before the PUT call, and the other in called after. Is it what you had in mind?

> Another problem I think of: if the same file is being used/linked by other resources, then deleting the file will make it unavailable for those other resources. Currently there is nothing that allows to do this (if you want to add the same image on two documents, you must upload it twice), but this is a possible scenario in the future, isn't it ? For example if we had some kind of digital assets management library like https://chonky.io/

As you said: "Currently there is nothing that allows to do this", so let's be pragmatic, and first fix this deletion issue.

If in the future, we may want to share files between resources (not sure to see the use cases though...), it will need more brainstorm to correctly implement it. In that case, maybe the logic of upload/deletion will have not to be handled by Semapps but by the end-user apps who will know when it's necessary to delete what resource I think.

**Review by srosset81 — APPROVED** [review]

**Comment by srosset81** [comment]
> As you said: "Currently there is nothing that allows to do this", so let's be pragmatic, and first fix this deletion issue.

I agree with that :-) Thanks for the fixes, I'll merge that now.

**Pull request merged**